### ISSUE TITLE
std.posix.accept does not return when interrupted. 

### ISSUE THREAD

**Issue by curuvar** [ziglang/zig#19445]
### Zig Version

0.12.0-dev.3352+95cb93944
aarch64 GNU/Linux

### Steps to Reproduce and Observed Behavior

I have some code that uses std.net.server.accept to accept connections.  If I send the process a SIGTERM (which I handle with my own sigaction) the accept does not return and the process remains blocked.  This makes it difficult to gracefully clean things up.

### Expected Behavior

the accept function should return an "Interrupted" error.

### Other Information

The actual problematic behavior is in the accept function in /std/posix.zig.  That function calls the system accept function in a loop whose whole purpose seem to be to cause this behavior.   

If this behavior is desired in some cases, perhaps a flag could be added to toggle it on or off.

**Comment by rootbeer** [comment]
TL;DR: I think this is a duplicate of #6193.

As I understand the current situation, almost all of the Zig POSIXy syscall wrappers retry on `EINTR`.  Not just `accept`.   I think the motivation is that EINTR is very annoying in 99% of code that invokes syscalls, so the hope is to remove that annoyance for Zig programs.

The cost of this approach is that blocking syscalls can become uninterruptible, so Zig programs will have to do something more drastic like close the blocked-on file descriptor (or kill the containing process and let the OS do the cleanup) when they want a blocked thread to give up on an operation.  Or use non-blocking APIs.  (Or perhaps Zig could support a timeout on blocking syscalls?)

For your use case, does closing the file descriptor work?

Zig could support a flag to say a syscall should be interruptible, but that gets complicated quickly.  Its either an extra parameter on every syscall (and every wrapper), or its a global/thread flag that would apply to all syscall invocations (but that would break any invocations not ready for an interruption).

In the short-term, as a work-around I believe you can link against glibc (or musl) to get the C Library's interruptible syscall semantics.

Personally, I'm skeptical of Zig trying to repair broken POSIX semantics like this.  So I suspect eventually Zig will be changed to return EINTR to callers (at least in the POSIXy APIs, but maybe not in the higher-level APIs?), but EINTR is a complicated and annoying behavior, so it seems worth a try.

See also #2425.

**Comment by marler8997** [comment]
Abstractions are never perfect, it's a tradeoff between generality and simplicity.  However, in this case the abstraction is opt-in. 
 Apps always have the option to use the system api's directly and build up their own abstractions when appropriate.

That being said, I don't see much downside to adding an `Options` struct in Zig's `accept` abstraction to customize this behavior, and, leave the door open for future options as well.

**Comment by GalaxySnail** [comment]
> As I understand the current situation, almost all of the Zig POSIXy syscall wrappers retry on `EINTR`. Not just `accept`. I think the motivation is that EINTR is very annoying in 99% of code that invokes syscalls, so the hope is to remove that annoyance for Zig programs.
> 
> The cost of this approach is that blocking syscalls can become uninterruptible, so Zig programs will have to do something more drastic like close the blocked-on file descriptor (or kill the containing process and let the OS do the cleanup) when they want a blocked thread to give up on an operation. Or use non-blocking APIs. (Or perhaps Zig could support a timeout on blocking syscalls?)

FWIW, Python wrappers also retry on `EINTR`, but they call the signal handler first, and only retry if the signal handler doesn't raise an exception. For details: [PEP 475](https://peps.python.org/pep-0475/). I think Zig can do something similar.

**Comment by curuvar** [comment]
I actually tried closing the stream, but that turns out to be worse.  The underlying system call returns with BADF which causes an "unreachable" panic.

As a temporary work around I'm calling std.posix.poll (which also loops on INTR, but does supply a timeout which allow me an escape from the loop) and only do the accept if there is a connection waiting.

I concur this issue is covered by #6193

**Comment by rootbeer** [comment]
I think the "EBADF causes unreachable" issue is a separate bug.  See #6389.  Generally patches motivated by real code, as in your case, are acceptable motivations to relax Zig's errno handling assumptions.  The EBADF can either be returned or could be mapped to Unexpected.  (E.g., see #3880.)

The [PEP 475](https://peps.python.org/pep-0475/) link is a solid description of the motivations for hiding and handling interruptions.  I'd argue something like that belongs in Zig's OS-agnostic infrastructure, and the "posix wrapper" layer shouldn't quite get that involved with dispatching signals, but it really depends on what level of fidelity Zig wants to take for this layer.

**Comment by marler8997** [comment]
> I think the "EBADF causes unreachable" issue is a separate bug

In this case this is exposing a bug in the program which is what we want.

> I actually tried closing the stream, but that turns out to be worse.

Once you close a file descriptor, that descriptor is no longer valid and attempting to invoke a system call with it is the equivalent to undefined behavior at the OS level.  That file descriptor could for example have already been re-used by opening a new file and now you're passing a different file to whatever system call you're invoking that you thought!  This means the app now has undefined behavior that's hidden behind a race condition, one of the worst kind of bugs an program can have, and why in this case Zig does whatever it can to help you catch it and discourage it's use.

I'm not sure what your code looks like, but, if you're wanting a listen socket to pop, you can call shutdown on it.  Or if you want to continue using a signal, you can call `std.os.system.accept` and handle `EINTR` yourself.

**Comment by andrewrk** [comment]
duplicate of #6193
